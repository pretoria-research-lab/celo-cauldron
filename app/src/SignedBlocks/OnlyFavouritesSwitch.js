import React from "react";
import Switch from "react-switch";
import PropTypes from "prop-types";

OnlyFavouritesSwitch.propTypes = () => { 
	return { 
		setOnlyFavourites: PropTypes.func,
		onlyFavourites: PropTypes.bool
	}; 
};

export default function OnlyFavouritesSwitch(props) {
 
	return (
		<label>
			<p>Favourites only</p>
			<Switch onColor="#35D07F" onChange={(checked, event, id) => props.setOnlyFavourites(checked, event)} checked={props.onlyFavourites} />
		</label>
	);
}