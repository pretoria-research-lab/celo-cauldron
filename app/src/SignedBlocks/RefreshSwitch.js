import React from "react";
import Switch from "react-switch";
import PropTypes from "prop-types";

RefreshSwitch.propTypes = () => { 
	return { 
    stayAtHead: PropTypes.func,
    checked: PropTypes.bool
	}; 
};

export default function RefreshSwitch(props) {
 
    return (
      <label>
        <p>Stay at head</p>
        <Switch onColor="#35D07F" onChange={(checked, event, id) => props.stayAtHead(checked, event)} checked={props.checked} />
      </label>
    );
}