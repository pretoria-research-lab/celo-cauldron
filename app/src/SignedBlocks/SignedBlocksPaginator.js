import React from "react";
import PropTypes from "prop-types";
import Pagination from "react-bootstrap/Pagination";
import "bootstrap/dist/css/bootstrap.css";
import "bootstrap";

SignedBlocksPaginator.propTypes = () => { 
	return { 
		changeSignatureMap: PropTypes.func,
		pageList: PropTypes.any,
		changeToFirstPage: PropTypes.func,
		changeToLastPage: PropTypes.func
	};
};

const nextPage = (pageList, changeSignatureMap) => {
	pageList.forEach((page,i) => {
		if(page.active && i < pageList.length-1){
			changeSignatureMap(pageList[i+1].atBlock);
		}
	});
};

const previousPage = (pageList, changeSignatureMap) => {
	pageList.forEach((page,i) => {
		if(page.active && i > 0){
			changeSignatureMap(pageList[i-1].atBlock);
		}
	});
};

export default function SignedBlocksPaginator(props) {

	const { pageList, changeSignatureMap, changeToLastPage, changeToFirstPage } = props;

	return (
		<div className="row">
			<div className="column" id="react-paginate">
				<Pagination>
					<Pagination.First onClick={() => changeToFirstPage()}/>
					<Pagination.Prev onClick={() => previousPage(pageList, changeSignatureMap)}/>
					{pageList.map((page, i) => 
						page.active ? <Pagination.Item key={i} active onClick={() => changeSignatureMap(page.atBlock)}>{page.atBlock}</Pagination.Item> 
							: <Pagination.Item key={i} onClick={() => changeSignatureMap(page.atBlock)}>{page.atBlock}</Pagination.Item>
					)}
					<Pagination.Next onClick={() => nextPage(pageList, changeSignatureMap)}/>
					<Pagination.Last onClick={() => changeToLastPage()}/>
				</Pagination>
			</div> 
		</div>
	);
}