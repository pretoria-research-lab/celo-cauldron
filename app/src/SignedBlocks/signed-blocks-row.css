#checkmark {
    max-height: 25px;
}

