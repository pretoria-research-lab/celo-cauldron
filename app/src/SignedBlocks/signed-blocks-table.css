.signed-blocks table {
  table-layout: fixed;
  width: 50%;
}

.signed-blocks td:nth-child(2),
th:nth-child(2) {
  width: 40ch;
}

.signed-blocks td:nth-child(3),
th:nth-child(3) {
  width: 40ch;
}

.signed-blocks thead tr th {
  position: sticky;
  top: 0;
}

.signed-blocks td,th {
  padding: 0.3rem;
}

.missedNumber {
  width: 12ch;
}

.signedNumber {
  width: 12ch;
}