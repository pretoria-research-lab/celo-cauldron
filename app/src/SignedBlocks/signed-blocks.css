.technical{
    font-family: "Jost"
}

.address {
    font-family: 'Courier New', Courier, monospace;
    font-size: smaller;
}

#loadingLogo {
    width: 150px;
}

#sortIcon {
    width: 20px;
}

.signed-blocks th {    
    font-family: "Jost";
    font-size: smaller;
    cursor: pointer;
}

.signed-blocks thead {
    background-color: var(--celo-canvas-faint-gold);
}

.blockMapInfo {
    /* padding: 0.20rem !important; */
    font-family: 'Courier New', Courier, monospace;
    font-size: 10px;
    vertical-align: middle !important;
    /* padding-bottom: 0; */
}

.signature-300 {
    padding: 0.05rem !important;
    /* font-size: 0px; */
}

.signature-200 {
    padding: 0.05rem !important;
    /* font-size: 0px; */
}

.signature-100 {
    padding: 0.15rem !important;
}

.missed {
    background-color: var(--celo-accent-red);
    color: var(--celo-accent-red);
}

.signed {
    background-color: var(--celo-green);
    color: var(--celo-green);
}

.other {
    background-color: var(--celo-accent-cyan);
    color: var(--celo-accent-cyan);
}

p {
    font-size: larger;
}

li {
    font-size:larger;
}

.nb {
    color: var(--celo-accent-blue);
    font-family: Jost;    
}

.text-align-left {
    text-align: left;
}

#react-paginate {
    margin: auto;
}

.signed-blocks-content {
    align-self: center;
    align-content: center;
    align-items: center;
    width: 92%;
}

.page-link {
    color: var(--celo-accent-blue);
}

.page-item.active .page-link {
    background-color:  var(--celo-accent-blue);
}

.jump {
    color: var(--celo-accent-blue);
    /* font-weight: bold
    ; */
}

.btn-secondary {
    border-radius: 7%;
    background-color: var(--celo-canvas-faint-gold);
    color: var(--celo-blue);
    border-color: var(--celo-blue);
}

.controls {
    align-items: center;
}

#favouriteIconTableHeader {
    width: 12px;
}

.headerRow {
    margin-left: auto;
    margin-right: auto;
}

ul {
    margin-bottom: 0rem;
}