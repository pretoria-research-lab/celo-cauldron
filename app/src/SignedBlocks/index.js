import React, {Component} from "react";
import PropTypes from "prop-types";
import { ToastContainer } from "react-toastify";
import * as toast from "../Common/Notification";
import "./signed-blocks.css";
import {API_CONFIG, SIGNED_BLOCKS_API_CONFIG} from "../Utils/config";
import {getCurrentBlockNumber, getCurrentEpochNumber} from "../Utils/utils";
import SignedBlocksService from "../SignedBlocksService";
import SignedBlocksTable from "./SignedBlocksTable";
import SignedBlocksHeader from "./SignedBlocksHeader";
import SignedBlocksPaginator from "./SignedBlocksPaginator";

import * as contractkit from "@celo/contractkit";

const signedBlocksAPI = new SignedBlocksService();
const refreshBlockNumberMillis = 100;
const highwatermarkRefreshMillis = 2000;
const baseScale = 100;

class SignedBlocks extends Component 
{

	static get propTypes() { 
		return { 
			network: PropTypes.any,
			lookback: PropTypes.any
		}; 
	}

	constructor(props){
		super(props);

		const stayAtHead = localStorage.getItem("stayAtHead") === "true" ? true : false;
		const onlyFavourites = localStorage.getItem("onlyFavourites") === "true" ? true : false;
		const scale = parseInt(localStorage.getItem("scale")) || this.props.lookback;
		console.log(`Retrieved previous settings, stayAtHead ${stayAtHead}, scale ${scale}, onlyFavourites ${onlyFavourites}`);

		this.state	=	{	pageList: [], 
			lookback: scale, 
			atBlock: -1, 
			signatures: [], 
			loading: true, 
			signedBlocksAPIConfig: {}, 
			remoteNodeConfig: {}, 
			blockNumber: 0, 
			processing: false,
			stayAtHead: stayAtHead,
			highwatermark: { atBlock: 0},
			epoch: -1,
			onlyFavourites: onlyFavourites,
			currentSortFunction : () => {}
		};
	}

	getCurrentBlockNumber = async () => {
		try{
			const nodeProvider = this.state.remoteNodeConfig.remoteNode;
			const blockNumber = await getCurrentBlockNumber(nodeProvider);
			const epochNumber = await getCurrentEpochNumber(nodeProvider, blockNumber);
			this.setState({blockNumber: blockNumber}, () => this.setState({epochNumber: epochNumber}));
		}catch(error){
			toast.notify("ERROR", this.props.network + " is not responding");
			this.setState({blockNumber: -1});
		}
	}

	getRemoteNodeConfig = (network) => {
		const remoteNodeConfig = API_CONFIG.filter((item) => item.network.toLowerCase() === network.toLowerCase())[0];
		console.log(`Using remote node configuration - ${JSON.stringify(remoteNodeConfig)}`);
		return remoteNodeConfig;
	}	

	getSignedBlocksAPIConfig = (network) => {
		const signedBlocksAPIConfig = SIGNED_BLOCKS_API_CONFIG.filter((item) => item.network.toLowerCase() === network.toLowerCase())[0];
		console.log(`Using signed blocks API configuration - ${JSON.stringify(signedBlocksAPIConfig)}`);
		return signedBlocksAPIConfig;
	}

	getHeadBlock = () => {
		return this.state.highwatermark.atBlock;
	}

	getHighwatermark = async () => {
		
		try{
			const previousHighwatermark = this.state.highwatermark;
			const highwatermark = (await signedBlocksAPI.getHighwatermark(this.state.signedBlocksAPIConfig)).data;
			
			// console.table(highwatermark.signers);
			
			this.setState({highwatermark}, async () => {
				
				if(this.state.stayAtHead && (previousHighwatermark.atBlock !== highwatermark.atBlock)) {

					console.log("Retrieved new block ...");

					const signers = this.state.signatures;
					let recalculatedSigners = [];					
					
					// Process signers in extant records
					signers.forEach((record, index) => {
						
						// Check if each signer exists in highwatermark
						const highwatermarkSigner = highwatermark.signers.filter((element, index) => element.signer === record.signer)[0];
												
						if(record.tickArray[0] === ".")							
							// If the first block in the tickArray was signed, reduce the signed count
							record.counts.signatures--;		
						else if (record.tickArray[0] === "~")
							// If it was not a signer, reduce that count
							record.counts.notASigner--;
						else
							// Otherwise reduce the missed count
							record.counts.missedSignatures--;

						if(highwatermarkSigner){
							if(highwatermarkSigner.signed===true) {
								// console.log(`New signed block found for extant signer ${record.signer}, signatures++`);
								record.counts.signatures++;
							}
							else{
								console.log(`New missed block found for extant signer ${record.signer}, missedSignatures++`);
								record.counts.missedSignatures++;
							}							

							// Slice array from first index (removes [0] element)
							record.tickArray = record.tickArray.slice(1);
							// Add new highwatermark to the array
							record.tickArray.push(highwatermarkSigner.signed ? "." : "✘");
						}
						else{
							console.log(`No new block found for extant signer ${record.signer}, notASigner++`);
							record.counts.notASigner++;
							record.tickArray = record.tickArray.slice(1);
							record.tickArray.push("~");
						}
						recalculatedSigners.push(record);
					});

					// Process signers in highwatermark
					highwatermark.signers.forEach((record, index) => {

						// Check if each signer exists in existing data
						const extantSigner = signers.filter((element, index) => element.signer === record.signer)[0];

						// If it doesn't, this highwatermark must be it's first signature
						if(!extantSigner){

							let newRecord = { 
								validator: record.validator,
								signer: record.signer,
								validatorName: record.validatorName,
								tickArray : [],
								counts: {
									signatures: (record.signed ? 1 : 0),
									missedSignatures: (record.signed ? 0: 1),
									notASigner: (this.state.scale - 1)
								}
							};
							console.log(`New signer found: ${record.signer}!`);

							// Pad tickArray with notASigner 
							for(let i=0;i<this.state.lookback-1;i++)
								newRecord.tickArray.push("~");
							
							newRecord.tickArray.push((record.signed ? "." : "✘"));
							recalculatedSigners.push(newRecord);
						}
					});
					
					this.enrichData(recalculatedSigners);

					this.setState({signatures: recalculatedSigners}, 
						() => this.setState({atBlock: highwatermark.atBlock},
							() => this.setState({pageList: this.createPageList()},
								() => this.state.currentSortFunction(this.state.currentSortDirection)
							)));
				}					
			});
		}
		catch(error){
			console.log("Highwatermark not found, likely still being written");
		}	
	}

	initialise = async () => {
	
		await this.getHighwatermark();
		const atBlock = this.getHeadBlock();

		const blockNumberIntervalID = setInterval(this.getCurrentBlockNumber, refreshBlockNumberMillis);
		this.setState({blockNumberIntervalID:blockNumberIntervalID});

		const highwatermarkIntervalID = setInterval(this.getHighwatermark, highwatermarkRefreshMillis);
		this.setState({highwatermarkIntervalID:highwatermarkIntervalID});
		
		this.setState({atBlock}, 
			() => this.setState({currentSortFunction: this.sortByMissedCount}, 
				() => {	const pageList = this.createPageList();
					this.setState({pageList}, () => this.getSignatures());
				}
			));
	}

	componentDidMount = () => {

		this.setState({remoteNodeConfig: this.getRemoteNodeConfig(this.props.network)}, 
			() => this.setState({signedBlocksAPIConfig: this.getSignedBlocksAPIConfig(this.props.network)}, 
				() => this.initialise()
			));		
	}

	componentWillUnmount = () => {
		clearInterval(this.state.blockNumberIntervalID);
		clearInterval(this.state.highwatermarkIntervalID);
	}

	getAccountName = async (address, atBlock) => {		
		let accountName = address;
		try{
			const kit = contractkit.newKit(this.state.remoteNodeConfig.remoteNode);
			const accounts = await kit.contracts.getAccounts();
			const validatorName = await accounts.getName(address);
			accountName = validatorName;
		}
		catch(e){
			toast.notify("WARN",`Error retrieving name for ${address}, ignoring`);
			console.log(e);
			accountName = address;		
		}
		return accountName;
	}

	getValidatorFromSigner = async (signerAddress, atBlock) => {
		let validatorAddress = "0x";
		try{
			const kit = contractkit.newKit(this.state.remoteNodeConfig.remoteNode);
			const accounts = await kit.contracts.getAccounts();
			validatorAddress = await accounts.signerToAccount(signerAddress);
		}
		catch(e){
			console.log(e);
			toast.notify("WARN",`Error retrieving validator for ${signerAddress}, ignoring`);	
		}
		return validatorAddress;
	}

	sortByValidatorName = (flip) => {
		const signatures = this.state.signatures;
		signatures.sort((x, y) => {
			let a = (x.validatorName ? x.validatorName : x.validatorAddress).toLowerCase(),	b = (y.validatorName ? y.validatorName : y.validatorAddress).toLowerCase();
			if(flip){a = (y.validatorName ? y.validatorName : y.validatorAddress).toLowerCase(); b = (x.validatorName ? x.validatorName : x.validatorAddress).toLowerCase();}
			return (a === b ? 0 : a > b ? 1 : -1);
		});
		this.setState({signatures});
		this.setState({currentSortDirection: flip}, () => this.setState({currentSortFunction: this.sortByValidatorName}));
	}
	sortBySignerAddress = (flip) => {
		const signatures = this.state.signatures;
		signatures.sort((x, y) => {			
			let a = x.signer.toLowerCase(),	b = y.signer.toLowerCase();
			if(flip){a = y.signer.toLowerCase(); b = x.signer.toLowerCase();}
			return (a === b ? 0 : a > b ? 1 : -1);
		});
		this.setState({signatures});
		this.setState({currentSortDirection: flip}, () => this.setState({currentSortFunction: this.sortBySignerAddress}));
	}
	sortBySignedCount = (flip) => {
		const signatures = this.state.signatures;
		signatures.sort((x, y) => {
			let a = x.counts.signatures, b = y.counts.signatures;
			if(flip){a = y.counts.signatures; b = x.counts.signatures;}
			return (a === b ? 0 : a > b ? 1 : -1);
		});
		this.setState({signatures});
		this.setState({currentSortDirection: flip}, () => this.setState({currentSortFunction: this.sortBySignerAddress}));
	}
	sortByMissedCount = (flip) => {
		const signatures = this.state.signatures;
		signatures.sort((x, y) => {
			let a = x.counts.missedSignatures, b = y.counts.missedSignatures;
			if(flip){ a = y.counts.missedSignatures; b = x.counts.missedSignatures;}
			return (a === b ? 0 : a > b ? 1 : -1);
		});
		this.setState({signatures});
		this.setState({currentSortDirection: flip}, () => this.setState({currentSortFunction: this.sortByMissedCount}));
	}

	getLinkForAddress = (address) => {
		return this.state.remoteNodeConfig.blockExplorer + "address/" + address;
	}

	getLinkForValidatorAddress = (address) => {
		return this.state.remoteNodeConfig.blockExplorer + "address/" + address + "/celo";
	}

	enrichData = (data) => {		
		data.forEach(element => {
			element["signerLink"] = this.getLinkForAddress(element.signer);
			element["validatorLink"] = this.getLinkForValidatorAddress(element.validatorAddress);
			element["favourite"] = localStorage.getItem(element.signer);
			
			if(element.favourite === null)
				element.favourite = false;
			else if(element.favourite === "true")
				element.favourite = true;
			else
				element.favourite = false;
		});
	}

	getSignatures = () => { 
		if(!this.state.atBlock){
			toast.notify("WARN","Waiting to receive current block number");
		}
		else{
			this.setState({loading:true}, async () => {
				try{
					const response = await signedBlocksAPI.getBlocks(
						this.state.signedBlocksAPIConfig, 
						this.state.atBlock-this.state.lookback+1, 
						this.state.atBlock, 
						(this.state.onlyFavourites ? this.state.signatures.filter((e,i) => e.favourite === true) : []));
						
					var data = response.data;
					
					toast.notify("INFO", "Retrieved signatures at block " + this.state.atBlock);	
						
					this.enrichData(data);
				
					this.setState({signatures: data}, () => {
						console.log(this.state);
						this.sortByMissedCount(true);
						this.setState({loading:false});}
					);
				}
				catch (error) {
					console.log(error);
					toast.notify("WARN","Error retrieving signatures, please refresh");
					const signatures = [];
					this.setState({signatures}, () => this.setState({loading:true}));
				}
			});
		}		
	}

	createPageList = () => {

		var pageList = [];
		let minPage = this.state.signedBlocksAPIConfig.firstBlock + this.state.lookback - baseScale; // quoroum was reached and database begins

		console.log("lookback = " + this.state.lookback);

		let maxPage = this.getHeadBlock(); // max page is always moving forward

		const listLength = this.state.signedBlocksAPIConfig.paginationListLength;
		const atBlock = this.state.atBlock;
		const lookback = this.state.lookback;
		let addedPrevious = false;
		let addedNext = false;

		console.log(`Creating pagelist with ${atBlock} as the active page ${minPage} as the first and ${maxPage} as the last page`);
		
		// Add earlier pages, only down to the minimum
		for(let i=(listLength/2);i>0;i--){
			if((atBlock-(i*lookback)) >= minPage){
				pageList.push(
					{   atBlock: (atBlock-(i*lookback)),
						active: false
					}        
				);
				addedPrevious = true;
			}
		}

		// Add pivot page
		pageList.push(
			{   atBlock: atBlock,
				active: true
			}
		);

		// Add later pages, only up to the maximum
		for(let i=1;i<=(listLength/2);i++){
			if((atBlock+(i*lookback))<=maxPage){
				pageList.push(
					{   atBlock: (atBlock+(i*lookback)),
						active: false
					}        
				);
				addedNext = true;
			}
		}
		
		// If we haven't filled up length of pages, add in remainder (on either side of the pivot)
		if(pageList.length < (listLength+1)){

			const iterations = ((listLength+1)-pageList.length); 

			for(let i=0; i < iterations; i++){
				if(addedPrevious){
					if(	(pageList[0].atBlock-lookback) >= minPage){
						const newPageList = [
							{   atBlock: (pageList[0].atBlock-lookback),
								active: false
							}, 
							...pageList ];
						pageList = newPageList;
					} 
				} else if (addedNext) {
					const newPageList = [
						...pageList, 	
						{   atBlock: (pageList[pageList.length-1].atBlock+lookback),
							active: false
						}];
					pageList = newPageList;
				}
			}
		}
		return pageList;
	}

	changeToFirstPage = () => {
		const firstPage = this.state.signedBlocksAPIConfig.firstBlock + this.state.lookback - baseScale;
		this.changePage(firstPage);
	}

	changeToLastPage = () => {
		this.changePage(this.getHeadBlock());
	}

	changePage = (atBlock) => {
		console.log(`Changing block map to atBlock ${atBlock} with range of ${this.state.lookback}`);
		this.setState({atBlock}, 
			async () => 
			{	const pageList = this.createPageList();
				this.setState({pageList}, () => this.getSignatures());
			}
		);
	}

	changeMapScale = (scale) => {
		
		let lookback;
		clearInterval(this.state.highwatermarkIntervalID);

		switch (scale) {
		case 0:
			lookback = 100;
			break;
		case 1:
			lookback = 200;
			break;
		case 2:
			lookback = 300;
			break;
		default:
			lookback = 100;
		}

		if(lookback !== this.state.lookback){
			this.setState({loading:true}, () => {
				toast.notify("INFO","Scale updated to " + lookback);
				localStorage.setItem("scale", lookback);
				this.setState({lookback}, () => {
					this.createPageList();
					this.changePage(this.state.atBlock);
					const highwatermarkIntervalID = setInterval(this.getHighwatermark, highwatermarkRefreshMillis);
					this.setState({highwatermarkIntervalID:highwatermarkIntervalID});
				});
			});
		}
	}

	setStayAtHead = async (stayAtHead) => {
		console.log("Stay at head =" + stayAtHead);
		if(this.state.stayAtHead !== stayAtHead){
			console.log("Changing stayAtHead to " + stayAtHead);
			localStorage.setItem("stayAtHead", stayAtHead);
			this.setState({stayAtHead}, () => {
				if(this.state.stayAtHead === true)
					this.changeToLastPage();
			});
		}
	}

	setOnlyFavourites = (onlyFavourites, event) => {

		const favourites = this.state.signatures.filter((e,i) => e.favourite === true) || [];
		if((favourites.length===0) && (onlyFavourites === true)){
			toast.notify("WARN", "No favourites have been set yet");
		}
		else{
			console.log("Only favourites =" + onlyFavourites);
			if(this.state.onlyFavourites !== onlyFavourites){
				console.log("Changing onlyFavourites to " + onlyFavourites);
				localStorage.setItem("onlyFavourites", onlyFavourites);
				this.setState({onlyFavourites});
			}
		}
	}

	toggleFavourite = (signer) => {

		const signerToToggle = this.state.signatures.filter((e,i) => e.signer === signer)[0];

		console.log("Signer to toggle is " + JSON.stringify(signerToToggle));

		if(signerToToggle.favourite === false){
			signerToToggle["favourite"] = true;
			console.log("Setting favourite for " + signer);
			localStorage.setItem(signer, true);
		}
		else {
			signerToToggle["favourite"] = false;
			console.log("Removing favourite for " + signer);
			localStorage.setItem(signer, false);
		}
		const signatures = [...this.state.signatures.filter((e,i) => e.signer !== signer), signerToToggle];
		this.setState({signatures});		
	}

	jumpToBlock = (atBlock) => {

		console.log("Jumping to block " + atBlock);

		const maxPage =  this.getHeadBlock();
		
		if(atBlock < (this.state.signedBlocksAPIConfig.firstBlock + this.state.lookback - baseScale))
			toast.notify("WARN","Block cannot be less than " + (this.state.signedBlocksAPIConfig.firstBlock + this.state.lookback - baseScale));
		else if (atBlock > maxPage)
			toast.notify("WARN","Block cannot be greater than " + maxPage);
		else{
			this.setStayAtHead(false);
			this.changePage(+atBlock);
		}	
	}

	render = () => {
		return ( 			
			<div className="container-fluid signed-blocks-content">   
				<SignedBlocksHeader
					jumpToBlock={this.jumpToBlock} 
					epochNumber={this.state.epochNumber}
					atBlock={this.state.atBlock}
					changeMapScale={this.changeMapScale}
					scale={this.state.lookback}
					blockNumber={this.state.blockNumber}
					config={this.state.remoteNodeConfig} network={this.props.network}
					setStayAtHead={this.setStayAtHead}
					stayAtHead={this.state.stayAtHead}
					setOnlyFavourites={this.setOnlyFavourites}
					onlyFavourites={this.state.onlyFavourites}
					highwatermark={this.state.highwatermark}				
				/>
				{this.state.loading ? "" : 
					<SignedBlocksPaginator 
						pageList={this.state.pageList} 
						changeToFirstPage={this.changeToFirstPage}
						changeToLastPage={this.changeToLastPage}
						changeSignatureMap={this.changePage}
					/>}
				<SignedBlocksTable 
					sortByValidatorName={this.sortByValidatorName} 
					sortBySignerAddress={this.sortBySignerAddress}
					sortBySignedCount={this.sortBySignedCount}
					sortByMissedCount={this.sortByMissedCount}
					toggleFavourite={this.toggleFavourite}
					config={this.state.remoteNodeConfig} 
					loading={this.state.loading} 
					lookback={this.state.lookback} 
					atBlock={this.state.atBlock} 
					blockNumber={this.state.blockNumber} 
					signatures={this.state.onlyFavourites ? this.state.signatures.filter((e,i) => e.favourite === true) : this.state.signatures}/>
				<ToastContainer autoClose={toast.DEFAULT_AUTOCLOSE}/>
				<hr />
			</div>
		);
	}  
}

export default SignedBlocks;