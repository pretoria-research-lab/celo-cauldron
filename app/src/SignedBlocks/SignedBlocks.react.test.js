import React from "react";
import SignedBlocks from "./index";
import renderer from "react-test-renderer";

const props = {
	network: "Baklava"
};

test("Signed Blocks component matches snapshot", () => {
	const component = renderer.create(
		<SignedBlocks {...props}/>,
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot(); 
});