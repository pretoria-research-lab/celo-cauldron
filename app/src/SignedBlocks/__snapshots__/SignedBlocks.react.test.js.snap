// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signed Blocks component matches snapshot 1`] = `
<div
  className="container-fluid signed-blocks-content"
>
  <div
    className="column centered"
  >
    <div
      className="row page-header"
    >
      <img
        alt="Developer's icon"
        className="celoIcons"
        src="04-Doers (light bg).png"
      />
      <h1
        className="mt"
      >
        Baklava Signed Blocks Map
      </h1>
    </div>
    <div
      className="row"
    >
      <h4>
        A visual map of validator signatures for the Baklava network. Data is updated every 100 blocks.
      </h4>
    </div>
    <div
      className="row jump controls"
    >
      <div
        className="col-sm-3"
      >
        <p>
          Current block 0
        </p>
      </div>
      <div
        className="col-sm-3"
      >
        <label>
          <p>
            Stay at head
          </p>
          <div
            style={
              Object {
                "MozTransition": "opacity 0.25s",
                "MozUserSelect": "none",
                "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
                "WebkitTransition": "opacity 0.25s",
                "WebkitUserSelect": "none",
                "borderRadius": 14,
                "direction": "ltr",
                "display": "inline-block",
                "msUserSelect": "none",
                "opacity": 1,
                "position": "relative",
                "textAlign": "left",
                "touchAction": "none",
                "transition": "opacity 0.25s",
                "userSelect": "none",
              }
            }
          >
            <div
              className="react-switch-bg"
              onClick={[Function]}
              onMouseDown={[Function]}
              style={
                Object {
                  "MozTransition": "background 0.25s",
                  "WebkitTransition": "background 0.25s",
                  "background": "#888888",
                  "borderRadius": 14,
                  "cursor": "pointer",
                  "height": 28,
                  "margin": 0,
                  "position": "relative",
                  "transition": "background 0.25s",
                  "width": 56,
                }
              }
            >
              <div
                style={
                  Object {
                    "MozTransition": "opacity 0.25s",
                    "WebkitTransition": "opacity 0.25s",
                    "height": 28,
                    "opacity": 0,
                    "pointerEvents": "none",
                    "position": "relative",
                    "transition": "opacity 0.25s",
                    "width": 30,
                  }
                }
              >
                <svg
                  height="100%"
                  style={
                    Object {
                      "position": "absolute",
                      "top": 0,
                    }
                  }
                  viewBox="-2 -5 17 21"
                  width="100%"
                >
                  <path
                    d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                    fill="#fff"
                    fillRule="evenodd"
                  />
                </svg>
              </div>
              <div
                style={
                  Object {
                    "MozTransition": "opacity 0.25s",
                    "WebkitTransition": "opacity 0.25s",
                    "height": 28,
                    "opacity": 1,
                    "pointerEvents": "none",
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "transition": "opacity 0.25s",
                    "width": 30,
                  }
                }
              >
                <svg
                  height="100%"
                  style={
                    Object {
                      "position": "absolute",
                      "top": 0,
                    }
                  }
                  viewBox="-2 -5 14 20"
                  width="100%"
                >
                  <path
                    d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                    fill="#fff"
                    fillRule="evenodd"
                  />
                </svg>
              </div>
            </div>
            <div
              className="react-switch-handle"
              onClick={[Function]}
              onMouseDown={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "MozTransition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
                  "WebkitTransition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
                  "background": "#ffffff",
                  "border": 0,
                  "borderRadius": "50%",
                  "boxShadow": null,
                  "cursor": "pointer",
                  "display": "inline-block",
                  "height": 26,
                  "outline": 0,
                  "position": "absolute",
                  "top": 1,
                  "transform": "translateX(1px)",
                  "transition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
                  "width": 26,
                }
              }
            />
            <input
              checked={false}
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyUp={[Function]}
              role="switch"
              style={
                Object {
                  "border": 0,
                  "clip": "rect(0 0 0 0)",
                  "height": 1,
                  "margin": -1,
                  "overflow": "hidden",
                  "padding": 0,
                  "position": "absolute",
                  "width": 1,
                }
              }
              type="checkbox"
            />
          </div>
        </label>
      </div>
      <div
        className="col-sm-3"
      >
        <div
          className="column"
        >
          <input
            className="form-control"
            defaultValue={0}
            id="jumpToBlock"
            name="jumpToBlock"
            onChange={[Function]}
            type="text"
          />
          <button
            className="btn-secondary"
            onClick={[Function]}
          >
            Jump to block
          </button>
        </div>
      </div>
      <div
        className="col-sm-3"
      >
        <p>
          Map Scale
        </p>
        <div
          className="rc-slider rc-slider-with-marks"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          onTouchStart={[Function]}
        >
          <div
            className="rc-slider-rail"
            style={Object {}}
          />
          <div
            className="rc-slider-track"
            style={
              Object {
                "left": "0%",
                "right": "auto",
                "width": "0%",
              }
            }
          />
          <div
            className="rc-slider-step"
          >
            <span
              className="rc-slider-dot rc-slider-dot-active"
              style={
                Object {
                  "left": "0%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "100%",
                }
              }
            />
          </div>
          <div
            aria-disabled={false}
            aria-valuemax={1}
            aria-valuemin={0}
            aria-valuenow={0}
            className="rc-slider-handle"
            onBlur={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="slider"
            style={
              Object {
                "left": "0%",
                "right": "auto",
                "transform": "translateX(-50%)",
              }
            }
            tabIndex={0}
          />
          <div
            className="rc-slider-mark"
          >
            <span
              className="rc-slider-mark-text rc-slider-mark-text-active"
              onMouseDown={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "color": "var(--celo-blue)",
                  "left": "0%",
                  "msTransform": "translateX(-50%)",
                  "transform": "translateX(-50%)",
                }
              }
            >
              <strong>
                100
              </strong>
            </span>
            <span
              className="rc-slider-mark-text"
              onMouseDown={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "color": "var(--celo-blue)",
                  "left": "100%",
                  "msTransform": "translateX(-50%)",
                  "transform": "translateX(-50%)",
                }
              }
            >
              <strong>
                1000
              </strong>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  
  <div
    className="row"
  >
    <div
      className="col-lg-12 text-center"
    >
      <img
        alt="Loading..."
        id="loadingLogo"
        src="loading.svg"
      />
    </div>
  </div>
  <div
    className="Toastify"
  />
  <hr />
</div>
`;
